### ISSUE TITLE
Implement CI/CD Features with GitHub Actions

### ISSUE THREAD

**Issue by rhughes42** [opensource-construction/website#83]
### Description:
We need to set up Continuous Integration (CI) and Continuous Deployment (CD) for our project to automate the build, test, and deployment processes. This will ensure our codebase remains stable and that new changes are deployed seamlessly. The CI/CD pipeline should be implemented using GitHub Actions.

#### Objectives:
1. **Set up Continuous Integration (CI) Pipeline**
   - Configure GitHub Actions to run on every push and pull request to the main branch.
   - Define steps to:
     - Install dependencies.
     - Run linting checks.
     - Execute unit tests.
     - Build the application.

2. **Set up Continuous Deployment (CD) Pipeline**
   - Configure GitHub Actions to deploy the application when changes are merged into the main branch.
   - Define deployment steps for staging and production environments.
   - Ensure deployment only occurs if CI checks pass.

3. **Add Notifications**
   - Configure notifications for CI/CD status (success/failure) to be sent to a Slack channel or via email.

#### Acceptance Criteria:
- CI pipeline runs automatically on every push and pull request to the main branch.
- CI pipeline includes steps for dependency installation, linting, testing, and building.
- CD pipeline deploys changes to the staging environment upon successful merge to the main branch.
- Deployment to production occurs only after manual approval.
- Notifications are sent for CI/CD status updates.

#### Tasks:
1. **CI Pipeline Configuration**
   - [ ] Create a GitHub Actions workflow file for CI.
   - [ ] Add steps for dependency installation (e.g., `npm install`).
   - [ ] Add steps for linting (e.g., `npm run lint`).
   - [ ] Add steps for running unit tests (e.g., `npm test`).
   - [ ] Add steps for building the application (e.g., `npm run build`).

2. **CD Pipeline Configuration**
   - [ ] Create a GitHub Actions workflow file for CD.
   - [ ] Add steps for deploying to the staging environment.
   - [ ] Add manual approval step for deploying to production.
   - [ ] Add steps for deploying to the production environment.

3. **Notifications Configuration**
   - [ ] Set up GitHub Actions to send CI/CD status notifications to a specified Slack channel or email.

#### Additional Context:
- Refer to the [GitHub Actions documentation](https://docs.github.com/en/actions) for detailed setup instructions.
- Ensure environment variables and secrets required for deployment are securely stored in GitHub Secrets.

#### Labels:
- enhancement
- CI/CD
- GitHub Actions

#### Assignees:
- [Team member names]

**Comment by n0rdlicht** [comment]
@rhughes42 many thanks for the well structured outline. I very much agree with the objectives and goals. To refine the initial tasks I'll add a bit of color to the current setup and context so we can focus on the gaps:

- adding a GitHub action for running lint on every push as well as ensuring the results are visible on PRs for reviewers
- adding a GitHub action for running unittests on every push as well as ensuring the results are visible on PRs for reviewers

Potential areas to investigate

- creating/adapting a custom discord chat bot to notify and receive/execute simple commands
- expanding and showing unit test coverage of component library

Let me know what you think and if the below is useful or needs clarification/follow-up.

## Current setup

### CI

next app
- `npm run lint`: currently runs eslint, no github action and/or pre-push hook setup
- unit testing - `npm run test`: currently runs `vitest --run`, no github action and/or pre-push hook setup, test are located in `/components/__tests__`
- `npm run build`: partially covered and automatically reported by Vercel build process running on every commit (see below)

components storybook
- visual testing -`.github/workflows/chromatic.yaml`: builds and uploads storybook to [chromatic](https://www.chromatic.com/docs/)

Current list of checks
<img width="865" alt="Screenshot 2024-06-21 at 08 58 35" src="https://github.com/opensource-construction/website/assets/1425577/18e5f348-95a6-4b9a-90de-0ad0aa15f811">
<img width="853" alt="Screenshot 2024-06-21 at 08 59 14" src="https://github.com/opensource-construction/website/assets/1425577/f2a39fca-0cc0-478e-8f71-ea7c2a558ebf">

### CD
summary: Vercel integration/hosting + basic `main` branch protection on GitHub

next app
- "deploy to staging": [Vercel public preview deployment](https://vercel.com/docs/deployments/preview-deployments) per branch/PR
   <img width="928" alt="Screenshot 2024-06-21 at 08 57 15" src="https://github.com/opensource-construction/website/assets/1425577/25c320f5-2de3-40f4-9bd5-aaf522b03662">
- approval: gh repo settings require web team review/approval before merging a PR
  <img width="928" alt="Screenshot 2024-06-21 at 08 57 58" src="https://github.com/opensource-construction/website/assets/1425577/d225e416-15d6-43ac-a9f9-5bcfb70360b5">
- "deploy to prod": Automatic [Vercel deployment](https://vercel.com/docs/deployments/git/vercel-for-github) from `main`

components storybook
- deployment - `.github/workflows/deploy-storybook-gh-pages.yaml`: deploy storybook on any push to `main` via GitHub pages to https://design.opensource.construction -> creates a check shown in PR and allows workflow to review and approve changes

### Notifications

Good point. I started looking into Discord bots, mainly for form submissions etc. Would be cool to have a custom os.c bot delivering notifications over there but also having a couple of commands for ChatOps. Regarding deployments I think the combination of review requirements, CI results and Vercel comments on PRs are good enough for now.

**Comment by rhughes42** [comment]
@n0rdlicht thanks for all the feedback. Do you mind adding me to the Vercel team? I just requested it this morning.
